src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt

```diff
@@ -230,6 +243,125 @@ internal val Class<*>.allDeclaredFieldWithSuperclasses get(): List<Field> {
+
+    val process: (
+        onRead: () -> Any?,
+        onCallback: (Any?) -> Boolean
+    ) -> Unit = { onRead, onCallback ->
+        // We wrap an unsafe read into `runCatching` to hande `UnsupportedOperationException`,
+        // which can be thrown, for instance, when attempting to read a field of
+        // a hidden class (starting from Java 15).
+        val result = runCatching { onRead() }
+
+        // do not pass fields to the user, that are non-readable by Unsafe
+        if (result.isSuccess) {
+            val fieldValue = result.getOrNull()
+
+            if (
+                onCallback(fieldValue) && // user determines whether to append to queue
+                fieldValue != null &&
+                !fieldValue.javaClass.isPrimitive && // no primitives traversing
+                !fieldValue.isPrimitiveWrapper && // no primitive wrappers traversing
+                !isImmutable(fieldValue) && // no immutable objects traversing
+                fieldValue !in visitedObjects // no reference-cycles allowed during traversing
+            ) {
+                queue.add(fieldValue)
+                visitedObjects.add(fieldValue)
+            }
+        }
+    }
+
+    while (queue.isNotEmpty()) {
+        val currentObj = queue.removeFirst()
+
+        when {
+            currentObj is Class<*> -> {}
+            currentObj.javaClass.isArray || isAtomicArray(currentObj) -> {
+                val length = getArrayLength(currentObj)
+                // TODO: casting currentObj to atomicfu class and accessing its field directly causes compilation error,
+                //  see https://youtrack.jetbrains.com/issue/KT-49792 and https://youtrack.jetbrains.com/issue/KT-47749
+                val cachedAtomicFUGetMethod: Method? = if (isAtomicFUArray(currentObj)) currentObj.javaClass.getMethod("get", Int::class.java) else null
+
+                for (index in 0..length - 1) {
+                    process(
+                        {
+                            if (isAtomicFUArray(currentObj)) {
+                                cachedAtomicFUGetMethod!!.invoke(currentObj, index) as Int
+                            }
+                            else when (currentObj) {
+                                is AtomicReferenceArray<*> -> currentObj.get(index)
+                                is AtomicIntegerArray -> currentObj.get(index)
+                                is AtomicLongArray -> currentObj.get(index)
+                                else -> readArrayElementViaUnsafe(currentObj, index)
+                            }
+                        },
+                        { onArrayElement(currentObj, index, it) }
+                    )
+                }
+            }
+            else -> {
+                // we jump through most of the atomic classes
+                var jumpObj: Any? = currentObj
+
+                if (isAtomic(jumpObj) && jumpObj !is AtomicReference<*> && jumpObj !is kotlinx.atomicfu.AtomicRef<*>) {
```

Comment:
- absence of `jumpObj !is AtomicReference<*>` causes [ObstructionFreedomViolationEventsCutTest](https://teamcity.jetbrains.com/buildConfiguration/KotlinTools_KotlinxLincheck_BuildAggregated/4868725?hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildDeploymentsSection=false&expandBuildChangesSection=true&expandBuildTestsSection=true&expandBuildProblemsSection=true) to fail
- `jumpObj !is kotlinx.atomicfu.AtomicRef<*>` added for symmetry

Reply:
Removed it after this [refactoring](https://github.com/JetBrains/lincheck/pull/420#discussion_r1826093690)